### ISSUE TITLE
Allow testing git-pile with different versions of git

### ISSUE THREAD

**Issue by guludo** [git-pile/git-pile#92]
We support git>=2.19. However it is possible that a contributor has a more recent version of git and could add code that is not backward-compatible with our supported version. An example is #87, which used an option for `git ls-tree` that was only available as of version 2.36 of git.

While it is fairly easy to download git's source, compile it and run the tests using `PATH="path/to/git/version/2.19/bin:$PATH" bats test`, it would be nice if we had something in our test infra-structure that made it easier to test with different git versions. Then, we could enable tests with git 2.19 and also with the latest stable release by default.

Some options in this direction:

1. There is a tool created specifically for managing versions of tools, called [asdf](https://asdf-vm.com/). It seems adding a plugin for git and using it here would not be that difficult.
2. Another option would be adding a home-made script that would download and and compile the two git versions for us.
3. Investigate using github action configuration to select the correct git version. That would make things not so easy to run tests locally, though.

In cases (1) and (2), we could then have a script that invoked the tests with each git version.

**Comment by lucasdemarchi** [comment]
I'd go with (3) since it seems much simpler. I don't want to use github actions to build versions of git - I want to support whatever users can have from reasonable recent versions of distros. For some definition of reasonable. If it's hard to support an old one, we can consider dropping it.

We already use build matrix in CI for python:
https://github.com/git-pile/git-pile/blob/master/.github/workflows/build.yml#L13

We should use something similar, installing a different version of git.  For reproducing things locally, it's pretty easy to create a container with whatever distro reproduces the issue and use git from there.

**Comment by guludo** [comment]
Well, I kinda saw your comments too late :-)

Here's a branch implementing option (2): https://github.com/git-pile/git-pile/compare/master...guludo:git-pile:test-git-versions

**Comment by lucasdemarchi** [comment]
I don't think we should spend CI time building git from its repo. It's not only a CI timing issue, but also that we should rather support git versions people are likely to be using. Most people get git from their distro (or a ppa, overlay, copr, etc)

**Comment by guludo** [comment]
Makes sense. However I think we also must be sure to support the minimal version we say to support. If we can guarantee that at least one of the distros to be included in CI has it, then great.

I also think it would be nice to be able to use the latest version in the tests for two reasons: (i) to make sure that we stop using things that are deprecated by git; (ii) when possible, to leverage any new feature that would improve git-pile.

For (i), I believe there might be cases when we will not be able to remove the deprecated thing entirely because of us supporting a minimal version, but we would at least have the new solution in the code in one of the branches of an ` if/else` statement.

Testing on the latest version is not really crucial for (ii), but would help making sure that the usage of the new features really work.

Regarding CI time, yes building them on every run would be not ideal. Maybe we could use the ppa, as you suggested. I checked and it contains the latest stable release as well as the minimal version we support. Not sure if there is an easy way of using git's official ppa on github actions (I don't have experience configuring github actions yet).

**Comment by guludo** [comment]
Why was this closed? Is it a wontfix?

**Comment by lucasdemarchi** [comment]
Oh, my mistake. I saw the PR closed and thought this was a dangling issue.
